### ISSUE TITLE
Implement malicious content monitoring for file attachments

### ISSUE THREAD

**Issue by yaelberger-commits** [cds-snc/notification-planning#725]
Checklist:
- [x] infrastructure: create a new S3 bucket and set up scan-files lambda function
- [x] dd-api changes pt 1: write a new endpoint that checks the malware scan verdict of file attachments
- [x] changes to notification-api to call the new endpoint in dd-api and refuse to send the email if malware is detected
- [x] add content to notification-admin that will display if the virus scan fails
- [x] extra safety check in dd-api: return an error code if a file containing malware is requested (@jzbahrai )
- [x] change to dd-api to protect Notify if scan-files goes down
- [x] finalize the ADR 
- [x] if scanning is not done, retry the send quickly the first time (rather than waiting 5 minutes)
- [x] alert in #notification-ops channel if malware is detected
- [x] dd-api should return a 408 if scan timed out (@smcmurtry)

QA:
- [x] send one malware, make sure it fails
- [x] send several malware files make alarms go off (5 warning 10 critical?)
- [x] send good file, make sure everything works
- [x] send with good and bad file, make sure it fails
- [x] send with 2 good files, make sure it works

**Comment by yaelberger-commits** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Notify-Planning-614b3ad91bc2030015ed22f5/issues/cds-snc/notification-planning/725?planning-poker) with ZenHub @andrewleith @jimleroyer @jzbahrai @sastels

**Comment by yaelberger-commits** [comment]
Maybe an ADR

**Comment by mohamed-cds** [comment]
With all the work @patheard and I have done with [scan-files](https://github.com/cds-snc/scan-files) this may be as simple as generating a `.html` of the template and feeding it to scan-files and getting a scan verdict. More investigation required but from what i can tell all the requirements can be met using the current scan-files api.

**Comment by jimleroyer** [comment]
It should be super easy to generate a HTML file for emails notification as this is what our markdown code will generate anyway.

**Comment by yaelberger-commits** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Notify-Planning-614b3ad91bc2030015ed22f5/issues/cds-snc/notification-planning/725?planning-poker) with Zenhub @smcmurtry @whabanks

**Comment by andrewleith** [comment]
- @smcmurtry had a chat with Max and has a good sense of where to go with this
- The first part of this will be to check uploaded files for malicious content (scan-files)

**Comment by smcmurtry** [comment]
We have a bunch of unused code that used to scan uploaded files for viruses. I will remove this in preparation for integrating the scan-files project. cc @jimleroyer

**Comment by jimleroyer** [comment]
Terraform PR is up, reviewed by Pat and Max and ready to be merged. Another part of the implementation in dd-api is being worked on to check file status.

**Comment by YedidaZalik** [comment]
@smcmurtry wrote some tests for this and one piece is ready to merge in Terraform - it's been reviewed and approved
@smcmurtry can merge after @jzbahrai deploys and then will continue with rest of work on this

**Comment by smcmurtry** [comment]
Update:
- merged the terraform changes to get file scanning working on our S3 bucket for file attachments
- discovered (with Pat’s help) that this approach will not work because these files are each encrypted with different keys
- started looking at a different approach for integrating scan-files: just calling their API directly

Next step:
- update my document-download-api PR to call the scan-files API directly, figure out if this is going to work

**Comment by smcmurtry** [comment]
Update:
- I have updated the document-download-api PR to call the scan-files API directly
- I have been able to test with small file attachments and everything works
- I still need to test the async behaviour, which I have been trying to do with larger file attachments that take longer to scan
- My API calls to the document-download-api are all hanging when I send large files there locally, even when I use the main branch that is working correctly in staging.
- I am trying to resolve that to unblock my testing

**Comment by andrewleith** [comment]
- @smcmurtry looking into the `flask-async` option today

**Comment by yaelberger-commits** [comment]
Start back on this today
Steve M will chat with Sharly and Pat about policy implications for options

**Comment by YedidaZalik** [comment]
Steve M is sick 
Was merged into staging yesterday and were hoping for Steve's ok before pushed to prod
This is high priority for this sprint
Steve A and Andrew can help

**Comment by smcmurtry** [comment]
Terraform changes are now released and working.
DD-api changes are approved and will be merged today.
notification-api changes are drafted but still need some work on tests. Steve M to work on that today.

**Comment by andrewleith** [comment]
Tested the following scenarios:
- [x] Linked file, clean
- [x] Linked file, malicious
- [x] Linked file, upload failure (delete from s3)
- [x] Attached file, clean
- [x] Attached file, malicious
- [x] Attached file, upload failure (delete from s3)

**Comment by smcmurtry** [comment]
Steve to review the dd-api PR.

**Comment by smcmurtry** [comment]
Jumana's PR on the safety check is merged. Andrew's alarm PR is merged. Steve M to work on the 408 error PR today, Jumana to review.

**Comment by andrewleith** [comment]
- All PRs are up / in review

**Comment by sastels** [comment]
and we did it! yay!
